peer/internal/uprng/uprng.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2024 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package uprng
+
+import (
+	cryptorand "crypto/rand"
+	"io"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/chacha20"
+)
+
+// Reader implements a cryptographically secure userspace PRNG that is
+// periodically reseeded with entropy obtained from crypto/rand.
+// Reader is safe for concurrent access.
+var Reader io.Reader
+
+func init() {
+	Reader = new(prng)
+}
+
+const (
+	maxCipherRead     = 4 * 1024 * 1024 // 4 MiB
+	maxCipherDuration = 20 * time.Second
+)
+
+var (
+	key   = make([]byte, chacha20.KeySize)
+	nonce = make([]byte, chacha20.NonceSize)
+)
+
+type prng struct {
+	cipher *chacha20.Cipher
+	read   int
+	t      time.Time
+	mu     sync.Mutex
+}
+
+func (p *prng) seed() {
+	_, err := cryptorand.Read(key)
+	if err != nil {
+		panic(err)
```

Comment:
Same here with the panics.  These are particularly notable because the only place `seed` is called is from `Read` which has an error return on it.

Reply:
See my ted talk above.  It's currently documented that the reader must never error.  So for this implementation, we don't return errors.